Content.Server/Decals/DecalSystem.cs

```diff
@@ -0,0 +1,227 @@
+            foreach (var (uid, decal) in ChunkCollections[gridId].GetChunk(chunkIndices))
+            {
+                if (tileIndices.X != (int) Math.Floor(decal.Coordinates.X) ||
+                    tileIndices.Y != (int) Math.Floor(decal.Coordinates.Y))
+                    continue;
+
+                if (validDelegate == null || validDelegate(decal))
+                {
+                    uids.Add(uid);
+                }
+            }
+
+            return uids;
+        }
+
+        public bool SetDecalPosition(uint uid, EntityCoordinates coordinates)
+        {
+            if (!ChunkIndex.TryGetValue(uid, out var values))
+            {
+                return false;
+            }
+
+            DirtyChunk(values);
+
+            var gridId = coordinates.GetGridId(EntityManager);
+            var decal = ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid];
+            if (gridId == values.gridId)
+            {
+                ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid] =
+                    decal.WithCoordinates(coordinates.Position);
+                return true;
+            }
+
+            ChunkCollections[values.gridId].GetChunk(values.chunkIndices).Remove(uid);
+            var chunkIndices = ChunkCollections[gridId].GetIndices(coordinates.Position);
+            ChunkCollections[gridId].GetChunk(chunkIndices).Add(uid, decal.WithCoordinates(coordinates.Position));
+            ChunkIndex[uid] = (gridId, chunkIndices);
+            DirtyChunk(gridId, chunkIndices);
+            return true;
+        }
+
+        public bool SetDecalColor(uint uid, Color? color)
+        {
+            if (!ChunkIndex.TryGetValue(uid, out var values))
+            {
+                return false;
+            }
+
+            var decal = ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid];
+            ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid] = decal.WithColor(color);
+            DirtyChunk(values);
+            return true;
+        }
+
+        public bool SetDecalId(uint uid, string id)
+        {
+            if (!ChunkIndex.TryGetValue(uid, out var values))
+            {
+                return false;
+            }
+
+            if (!PrototypeManager.HasIndex<DecalPrototype>(id))
+                throw new ArgumentOutOfRangeException($"Tried to set decal id to invalid prototypeid: {id}");
+
+            var decal = ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid];
+            ChunkCollections[values.gridId].GetChunk(values.chunkIndices)[uid] = decal.WithId(id);
+            DirtyChunk(values);
+            return true;
+        }
+
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            if (_removedDecals.Count > 0)
+                RaiseNetworkEvent(new DecalRemovalUpdateEvent{RemovedDecals = _removedDecals});
```

Comment:
Probably a bad idea for bandwidth reasons?

Reply:
how come? i guess i could tie it into the other updatemessage

Reply:
Say decals get changed on salvage; someone who's on the station doesn't probably care about whether they're removed or not as they've never seen it. Though I guess the trade off is that you need to track what chunks each player has seen.

Reply:
> Though I guess the trade off is that you need to track what chunks each player has seen.

not really. if i were to send this in the updatemessage, we would update all seen chunks by the player anyways. we dont really care if the data outside the viewport is incorrect